File: algorithms/c++/implementation/migratory_birds.cpp
#include <iostream>
#include <map>
#include<algorithm>

using namespace std;

int migratoryBirds(map<int,int> m) {
  int biggest = 0;
  int smallestKey = 0;
  for(auto const& item: m){
    if(item.second>biggest){
        biggest=item.second;
        smallestKey = item.first;
    }
  }
  return smallestKey;
}

int main() {
    int n;
    cin >> n;
    map<int,int> m;
    int v;
    for(int ar_i = 0; ar_i < n; ar_i++){
       cin >> v;
       if (m.count(v)>0)
            m[v] = m[v] + 1;
          else
            m[v] = 0;

    }
    int result = migratoryBirds(m);
    cout << result << endl;
    return 0;
}

File: algorithms/java/src/implementation/MigratoryBirds.java
package implementation;

import java.util.*;

public class MigratoryBirds {

    private static int migratoryBirds(TreeMap<Integer,Integer> birdsObservations) {
        int biggest = 0;
        int smallestKey = 0;
        for(Map.Entry<Integer,Integer> entry: birdsObservations.entrySet()){
            if(entry.getValue()>biggest){
                biggest = entry.getValue();
                smallestKey = entry.getKey();
            }
        }
        return smallestKey;
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        TreeMap<Integer,Integer> birdsObservations = new TreeMap<>();
        for(int ar_i = 0; ar_i < n; ar_i++) {
            int birdID = in.nextInt();
            if (birdsObservations.containsKey(birdID))
                birdsObservations.put(birdID, birdsObservations.get(birdID) + 1);
            else
                birdsObservations.put(birdID, 0);
        }
        int result = migratoryBirds(birdsObservations);
        System.out.println(result);
    }
}
